### PR TITLE
[RHPAM-3681] Add metering labels for Kogito resources

### PR DESCRIPTION
**Pull request by Kevin-Mok**
See: https://issues.redhat.com/browse/RHPAM-3681

Cherry-pick for `1.5.x` branch: https://github.com/kiegroup/rhpam-kogito-operator/pull/49

## Testing
I deployed this YAML:
```yaml
apiVersion: rhpam.kiegroup.org/v1
kind: KogitoRuntime
metadata:
  name: qe
spec:
  replicas: 1
  image: quay.io/kmok/process-quarkus-example
```
Where the resulting pod had the expected labels:
```
➜ kubectl describe pod qe-*
...
Labels: app=qe
        com.company=Red_Hat
        rht.comp=PAM
        rht.comp_ver=7.10
        rht.prod_name=Red_Hat_Process_Automation
        rht.prod_ver=7.11
        rht.subcomp=rhpam-kogito-runtime
        rht.subcomp_t=application
...
```
I did not write a unit test for the helper functions I wrote as I wasn't sure where and how tests are written in the product operator. Perhaps @sutaakar could direct me in how to write tests for this PR?

## Requirements
Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors' guide](https://github.com/kiegroup/kogito-cloud-operator/blob/master/README.md#contributing-to-the-kogito-operator)
- [x] Pull Request title is properly formatted: 
- [x] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [ ] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster
- [x] You've added a [RELEASE_NOTES.md](RELEASE_NOTES.md) entry regarding this change

### CODE REVIEW

File: version/version.go [hunk 1]
```diff
@@ -17,4 +17,6 @@ package version
 var (
 	// Version of Kogito Operator
 	Version = "7.11.0"
+	// PAMVersion used for metering labels
+	PAMVersion = "7.10"
```
Comment (ricardozanini) [inline version/version.go:21]: Why 7.10 and not 7.11? :thinking:

Comment (Kevin-Mok) [reply version/version.go:21]: I was following @ibek's comment in [the Jira](https://issues.redhat.com/browse/RHPAM-3681?focusedCommentId=16364445&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16364445).

**Comment by sutaakar** [comment]
@Kevin-Mok Can you please apply proper labels also to KogitoBuild pods?

File: version/version.go (hunk 1 above)
Comment (spolti) [reply version/version.go:21]: I think it should be 7.12.0 at this point. or, 7.11.1;

**Comment by spolti** [comment]
Good point @sutaakar 
@vaibhavjainwiz can you please review and see my conversation with Kevin on our cloud channel as well?

File: version/version.go (hunk 1 above)
Comment (sutaakar) [reply version/version.go:21]: The version updates are done in `bump-version.sh` so I think we can put 7.11.0 here, will be updated before release to proper version together with other places.

**Comment by Kevin-Mok** [comment]
I'm running into a couple of problems trying to add labels to the builder pods.

## Can't get a build pod running
I deployed the [`example/process-quarkus-example.yaml`](https://github.com/kiegroup/rhpam-kogito-operator/blob/master/example/process-quarkus-example.yaml) but just continually got this error:
```
ERROR   controller      Reconciler error
{"reconcilerGroup": "rhpam.kiegroup.org", "reconcilerKind": "KogitoRuntime", 
"controller": "kogitoruntime", "name": "example-quarkus", "namespace": "kogito", "error": "image not found"}
```
I then tried deploying only the KogitoRuntime but no build pod was created either. Here are some of the other resources in the namespace:
```
> oc get bc
NAME                              TYPE     FROM   LATEST
example-quarkus                   Source          0
example-quarkus-builder           Source   Git    0

> oc get builds
No resources found in kogito namespace.

> oc get is
NAME                             IMAGE REPOSITORY                                                                                TAGS       UPDATED
example-quarkus                  default-route-openshift-image-registry.apps-crc.testing/kogito/example-quarkus
example-quarkus-builder          default-route-openshift-image-registry.apps-crc.testing/kogito/example-quarkus-builder
```

## Where to add labels to build pods
The only relevant build file I found was the `controllers/kogitobuild_controller.go`, but there, I only see the `instance` object to work with. I can't modify the `Spec.Template.Labels` directly as I did with the runtime deployment; instead, I have to use the `GetSpec()` function where I can't find a relevant function in the [KogitoBuildSpecInterface](https://github.com/kiegroup/kogito-operator/blob/33e6b074674ee40d65f6db8b5b8f4e012bb0ddee/api/kogitobuild_types.go#L91-L122) to set the labels.

I see that I can also use `SetLabels()` on the `instance` directly, but when I tried to do this for the runtime, I got an error about trying to overwrite a selector after it's already been created. I was going to try to see if this was possible anyway, but I ran into the previous problem of trying to get a builder pod running.

**Comment by sutaakar** [comment]
@Kevin-Mok I guess you run the operator using `make run`, right? In such case the operator doesn't have proper environment variables specifying default image names to be used for build. As a workaround you can specify build and runtime images as part of the KogitoBuild CR.

**Comment by ricardozanini** [comment]
@Kevin-Mok @sutaakar in this scenario, I believe we need to either:

- Add a comprehensive message in the resource status
- Create a WARN event in the cluster

It's pretty difficult to debug what's going on otherwise. wdyt?

File: internal/labels.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	if deploymentLabels == nil {
+		deploymentLabels = map[string]string{}
+	}
+	for labelKey, labelValue := range meteringLabels {
+		deploymentLabels[labelKey] = labelValue
+	}
+	return deploymentLabels
+}
+
+func trimMinorVersion(version string) string {
```
Comment (vaibhavjainwiz) [inline internal/labels.go:45]: reuse below fcuntion in place of `trimMinorVersion`.
https://github.com/kiegroup/kogito-operator/blob/c03c9fd08199735d802a579eb4ca8cfc3c186db8/core/infrastructure/image.go#L166

File: internal/labels.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	"rht.prod_name": "Red_Hat_Process_Automation",
+	"rht.prod_ver":  trimMinorVersion(version.Version),
+	"rht.comp":      "PAM",
+	"rht.comp_ver":  version.PAMVersion,
+	"rht.subcomp":   "rhpam-kogito-runtime",
+	"rht.subcomp_t": "application",
+}
+
+// SetMeteringLabels for product operator
+func SetMeteringLabels(deploymentLabels map[string]string) map[string]string {
```
Comment (vaibhavjainwiz) [inline internal/labels.go:35]: reuse `AppendToStringMap` function here.
https://github.com/kiegroup/kogito-operator/blob/8ce30eefd84610f843e1207f5dd8fa00c1c12592/core/framework/util/maps.go#L56

**Comment by sutaakar** [comment]
Message in the status would be good.

**Comment by Kevin-Mok** [comment]
With help from @sutaakar, I was eventually able to get the build pod running with this YAML:
```yaml
apiVersion: rhpam.kiegroup.org/v1
kind: KogitoBuild
metadata:
  name: process-quarkus-example
spec:
  buildImage: registry.redhat.io/rhpam-7/rhpam-kogito-builder-rhel8:7.11.0
  gitSource:
    contextDir: process-quarkus-example
    uri: https://github.com/kiegroup/kogito-examples
  runtime: quarkus
  runtimeImage: registry.redhat.io/rhpam-7/rhpam-kogito-runtime-jvm-rhel8:7.11.0
  type: RemoteSource
```
If this was only an issue related to running the operator locally, then I would say this was a pretty non-intuitive solution. The [`example/process-quarkus-example-runtime-build.yaml`](https://github.com/kiegroup/rhpam-kogito-operator/blob/master/example/process-quarkus-example-runtime-build.yaml) shows that you need a `[build|runtime]Image`, but I would have never guessed they needed to be `registry.redhat.io/rhpam-7/rhpam-kogito-[builder|runtime]-jvm-rhel8:7.11.0`.

@ricardozanini, I can't speak too much on the error handling of this on the K8s side, but from a UX/community developer perspective, there seems like there could be some improvement here. Perhaps at least changing our `KogitoBuild` examples to include the `registry.redhat.io/rhpam-7/rhpam-kogito-*` images and updating the versioning in `hack/bump-version.sh`?

**Comment by vaibhavjainwiz** [comment]
> @Kevin-Mok Can you please apply proper labels also to KogitoBuild pods?

Kogito Operator doesn't directly create pods for KogitoBuild. When the build is triggered then these pods are controlled by other controller. I believe we should set metering label in `BuildConfig` resource and they should propogate to build pods by itself.

**Comment by ricardozanini** [comment]
> @ricardozanini, I can't speak too much on the error handling of this on the K8s side, but from a UX/community developer perspective, there seems like there could be some improvement here. Perhaps at least changing our `KogitoBuild` examples to include the `registry.redhat.io/rhpam-7/rhpam-kogito-*` images and updating the versioning in `hack/bump-version.sh`?

@vaibhavjainwiz @sutaakar, aren't we doing this already by adding env vars to the operator deployment? @Kevin-Mok, since you're running locally, the images won't be there. Although I would expect that if the envs are empty, the operator would assume defaults and, like you said, upgraded during the version bump.

**Comment by sutaakar** [comment]
>     @vaibhavjainwiz @sutaakar, aren't we doing this already by adding env vars to the operator deployment?

Yes, that should be done in case operator is deployed using OLM or YAML.

File: version/version.go (hunk 1 above)
Comment (spolti) [reply version/version.go:21]: I don't think we need to add this env as well, component version should be the same than the pro version.

File: internal/labels.go (hunk 3 above)
Comment (spolti) [reply internal/labels.go:35]: please add some unit tests to test if the kogito deployments and builds will have the expected labels.

**Review by spolti — CHANGES_REQUESTED** [review]

**Comment by Kevin-Mok** [comment]
> Kogito Operator doesn't directly create pods for KogitoBuild. When the build is triggered then these pods are controlled by other controller. I believe we should set metering label in `BuildConfig` resource and they should propogate to build pods by itself.

@vaibhavjainwiz @sutaakar I tested using the [`BuildContext` wrapper](https://github.com/vaibhavjainwiz/kogito-operator/commit/8449b389c0dce1f0382ed5a0362a5c89e51e8b93) @vaibhavjainwiz wrote in my [`RHPAM-3681-buildctx`](https://github.com/Kevin-Mok/rhpam-kogito-operator/tree/RHPAM-3681-buildctx) branch, but the labels passed to the `BuildConfig` are not passed to the build pods.
```
> oc describe bc process-quarkus-example[-builder] 
...
Labels: app=process-quarkus-example
        buildType=RemoteSource
        com.company=Red_Hat
        rht.comp=PAM
        rht.comp_ver=7.11
        rht.prod_name=Red_Hat_Process_Automation
        rht.prod_ver=7.11
        rht.subcomp=rhpam-kogito-runtime
        rht.subcomp_t=application
...

> oc describe pod process-quarkus-example-builder-1-build
...
Labels:       openshift.io/build.name=process-quarkus-example-builder-1
...
```

**Comment by sutaakar** [comment]
@Kevin-Mok ok, in such case it is IMHO ok to keep the labels just for runtime.

File: RELEASE_NOTES.md [hunk 4]
```diff
@@ -1,11 +1,13 @@
-The RHPAM Kogito Operator is based on Kogito Operator. 
-Please, see the upstream [RELEASE NOTES](https://github.com/kiegroup/kogito-operator/blob/master/RELEASE_NOTES.md).
-
-<!-- Remove the comment for any RHPAM JIRA exclusive 
-## Enhancements  
+The RHPAM Kogito Operator is based on Kogito Operator. Please see the upstream [RELEASE NOTES](https://github.com/kiegroup/kogito-operator/blob/master/RELEASE_NOTES.md).
 
 ## Bug Fixes
 - [KOGITO-4894](https://issues.redhat.com/browse/KOGITO-4894) - Remove `image` default value in KogitoRuntime CR (problem in OCP console)
 - [KOGITO-5104](https://issues.redhat.com/browse/KOGITO-5104) - Kogito operator creates wrong default image tags
```
Comment (sutaakar) [inline RELEASE_NOTES.md:5]: Those JIRAs don't seem to be related any more with next release, could you please remove them?

File: example/process-quarkus-example-runtime-kmok.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+apiVersion: rhpam.kiegroup.org/v1
```
Comment (sutaakar) [inline example/process-quarkus-example-runtime-kmok.yaml:1]: A debugging leftover?

**Review by sutaakar — APPROVED** [review]
Minor notes above, otherwise looks good.

**Comment by Kevin-Mok** [comment]
I've resolved all the review comments and addressed adding labels to the operator/builder pods and unit tests in [this Jira comment](https://issues.redhat.com/browse/RHPAM-3681?focusedCommentId=16386947&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16386947).

**Review by spolti — APPROVED** [review]

**Comment by spolti** [comment]
@Kevin-Mok, please cherry pick it on the 1.5.x branch as well.

**Review by ricardozanini — APPROVED** [review]

**Comment by Kevin-Mok** [comment]
If I'm using @vaibhavjainwiz's [`BuildContext` PR](https://github.com/kiegroup/kogito-operator/pull/917) to `kogito-operator@master` to update this PR, am I supposed to run `go get github.com/kiegroup/kogito-operator@master`? If so, I'm getting this error:
```
go get: github.com/kiegroup/kogito-operator@v0.12.1-0.20210531140744-03fa70509895 updating to
	github.com/kiegroup/kogito-operator@v0.12.1-0.20210706132755-6a8e795c9754 requires
	github.com/kiegroup/kogito-operator/api@v0.0.0-00010101000000-000000000000: invalid version: unknown revision 000000000000
```

Is this related to @ricardozanini's [API PR](https://github.com/kiegroup/kogito-operator/pull/916), or am I doing something wrong?

**Comment by ricardozanini** [comment]
@Kevin-Mok try adding this to the `go.mod` file:

```
replace (
        github.com/kiegroup/kogito-operator => github.com/kiegroup/kogito-operator master
	github.com/kiegroup/kogito-operator/api => github.com/kiegroup/kogito-operator/api master
	github.com/kiegroup/kogito-operator/client => github.com/kiegroup/kogito-operator/client master
)
```

Must be updated to the target Kogito Operator version once you do the product release. This is because they are local modules in the kogito-operator module.

**Comment by Kevin-Mok** [comment]
@ricardozanini Worked like a charm, thanks. :)

I updated the PR as per the [1.5.x PR](https://github.com/kiegroup/rhpam-kogito-operator/pull/49#issuecomment-874338774) so that the labels are added to the `Build` and `BuildConfig` resources.  I'm having some authentication issues with my `KogitoBuild`, so I tested the PR using a separate `KogitoBuild` and `KogitoRuntime`:
```yaml
apiVersion: rhpam.kiegroup.org/v1
kind: KogitoBuild
metadata:
  name: process-quarkus-example
spec:
  buildImage: registry.redhat.io/rhpam-7/rhpam-kogito-builder-rhel8:7.11.0
  gitSource:
    contextDir: process-quarkus-example
    uri: https://github.com/kiegroup/kogito-examples
  runtime: quarkus
  runtimeImage: registry.redhat.io/rhpam-7/rhpam-kogito-runtime-jvm-rhel8:7.11.0
  type: RemoteSource
---
apiVersion: rhpam.kiegroup.org/v1
kind: KogitoRuntime
metadata:
  name: qe
spec:
  replicas: 1
  image: quay.io/kmok/process-quarkus-example
```
Which then results in:
```
> oc get all --selector='com.company=Red_Hat'
NAME                     READY   STATUS             RESTARTS   AGE
pod/qe-f76fcc644-t8fc9   0/1     CrashLoopBackOff   0          9m45s
 
NAME                 READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/qe   0/1     1            0           13m
 
NAME                           DESIRED   CURRENT   READY   AGE
replicaset.apps/qe-f76fcc644   1         1         0       9m45s
 
NAME                                                             TYPE     FROM   LATEST
buildconfig.build.openshift.io/process-quarkus-example           Source          0
buildconfig.build.openshift.io/process-quarkus-example-builder   Source   Git    1
 
NAME                                                         TYPE     FROM          STATUS    STARTED         DURATION
build.build.openshift.io/process-quarkus-example-builder-1   Source   Git@9b97738   Running   7 minutes ago
```

**Review by spolti — APPROVED** [review]
